src/tslint.ts

```diff
@@ -42,17 +44,46 @@ class Linter {
 
     private fileName: string;
     private source: string;
+    private program: ts.Program;
     private options: ILinterOptions;
 
-    constructor(fileName: string, source: string, options: ILinterOptionsRaw) {
+    public static getProgram(configFile: string, projectDirectory?: string): ts.Program {
+        if (projectDirectory === undefined) {
+            const lastSeparator = configFile.lastIndexOf("/");
+            if (lastSeparator < 0) {
+                projectDirectory = ".";
+            } else {
+                projectDirectory = configFile.substring(0, lastSeparator + 1);
+            }
+        }
+
+        const {config} = ts.readConfigFile(configFile, ts.sys.readFile);
+        const parsed = ts.parseJsonConfigFileContent(config, {readDirectory: ts.sys.readDirectory}, projectDirectory);
+        const host = ts.createCompilerHost(parsed.options, true);
+        const program = ts.createProgram(parsed.fileNames, parsed.options, host);
```

Comment:
we need the program to be type-checked before it is handed to `applyWithProgram`. We could do the type-checking here, but then we have to report type-check errors which is awkward.
Alternatively we can leave the API user to do it, but we should verify that it was done, see https://github.com/angular/tsickle/commit/5d9281420b02121de788a41ecb8524d981744eb5 for an example